### PR TITLE
[EC-76] EVM Stop, Arithmetic, Comparison and Bitwise opcodes

### PR DESCRIPTION
**Pull request by whysoserious**
Introduces all missing `0`s and `10`s opcodes plus a few minor formatting changes.

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/vm/OpCode.scala [hunk 1]
```diff
@@ -228,16 +249,30 @@ case object EXP extends BinaryOp(0x0a, G_exp)(_ ** _) {
...
+
 case object EQ extends BinaryOp(0x14, G_verylow)((a, b) => DataWord(a == b)) with ConstGas
 
 case object ISZERO extends UnaryOp(0x15, G_verylow)(a => DataWord(a.isZero)) with ConstGas
 
 case object AND extends BinaryOp(0x16, G_verylow)(_ & _) with ConstGas
 
+case object OR extends BinaryOp(0x17, G_verylow)(_ & _) with ConstGas
+
+case object XOR extends BinaryOp(0x18, G_verylow)(_ & _) with ConstGas
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: Shouldn't the `XOR` function be `(_ ^ _)`? Same with `OR` and `|`.

File: src/main/scala/io/iohk/ethereum/vm/DataWord.scala [hunk 2]
```diff
@@ -79,10 +88,56 @@ class DataWord private (private val n: BigInt) extends Ordered[DataWord] {
...
+
+  def sgt(that: DataWord): DataWord = DataWord(this.signedN > that.signedN)
+
+  def getByte(that: DataWord): DataWord = {
+    val cnt: Byte = that.n.toByte
+    if (cnt < 0 || cnt > 31) {
+      Zero
+    } else {
+      DataWord(Zeros.take(Size - 1) :+ this.bytes(cnt))
+    }
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: While trying to undertand the BYTE opcode, I came up with the test: `DataWord(1).getByte(DataWord(287))` that currently returns `DataWord(1)`. As `287>=32`, shouldn't this test return `DataWord(0)` according to the paper? Maybe I'm thinking something wrong.

Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: Yes, the `.toByte` conversion is wrong.

File: src/test/scala/io/iohk/ethereum/vm/DataWordSpec.scala [hunk 3]
```diff
@@ -114,6 +168,57 @@ class DataWordSpec extends FunSuite with PropertyChecks {
     }
   }
 
+  test("signExtend") {
+    assert(DataWord(42).signExtend(DataWord(3)) == DataWord(42))
```
Comment (rtkaczyk) [inline src/test/scala/io/iohk/ethereum/vm/DataWordSpec.scala]: How about using a `Table` for such tests?

File: src/main/scala/io/iohk/ethereum/vm/DataWord.scala [hunk 4]
```diff
@@ -79,10 +88,56 @@ class DataWord private (private val n: BigInt) extends Ordered[DataWord] {
...
+      this
+    } else {
+      val mask: Int = if (this.signedN.testBit((cnt * 8) + 7)) 0xFF else 0x00
+      val leftFill: Array[Byte] = Array.fill(Size - cnt - 1)(mask.toByte)
+      val bs = ByteString(leftFill) ++ this.bytes.takeRight(cnt + 1)
+      DataWord(bs)
+    }
+  }
+
+  def slt(that: DataWord): DataWord = DataWord(this.signedN < that.signedN)
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: I'm not fond of the inconsistency in that normal comparators (`<`, `>`) return a `Boolean` whereas these (`sgt`, `slt`) return a `DataWord`.

File: src/main/scala/io/iohk/ethereum/vm/DataWord.scala [hunk 5]
```diff
@@ -79,10 +88,56 @@ class DataWord private (private val n: BigInt) extends Ordered[DataWord] {
 
   def *(that: DataWord): DataWord = DataWord(this.n * that.n)
 
-  def /(that: DataWord): DataWord = DataWord(this.n / that.n)
+  def /(that: DataWord): DataWord = zeroCheck(that) { DataWord(this.n / that.n) }
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: When you see a `/` method on a numeric type you would expect it to mean the usual division. It's not the case so it would make sense add a comment. But I think it's even better to rename it to `div`.

File: src/test/scala/io/iohk/ethereum/vm/DataWordSpec.scala [hunk 6]
```diff
@@ -114,6 +168,57 @@ class DataWordSpec extends FunSuite with PropertyChecks {
...
+    assert(DataWord(42).signExtend(DataWord(-1)) == DataWord(42))
+    assert(DataWord(42).signExtend(DataWord(0)) == DataWord(42))
+    assert(DataWord(42).signExtend(DataWord(Size)) == DataWord(42))
+    assert(DataWord(42).signExtend(DataWord(Size + 1)) == DataWord(42))
+    assert(DataWord(-42).signExtend(DataWord(Size)) == DataWord(-42))
+    assert(DataWord(-42).signExtend(DataWord(Size + 1)) == DataWord(-42))
+    assert(DataWord(-42).signExtend(DataWord(-11)) == DataWord(-42))
+    assert(DataWord(-1).signExtend(DataWord(1)) == DataWord(-1))
+    assert(DataWord(-1).signExtend(DataWord(1)) == DataWord(-1))
+    // 1737215.signExtend(1) == 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff81ff
```
Comment (rtkaczyk) [inline src/test/scala/io/iohk/ethereum/vm/DataWordSpec.scala]: Commented out code

**Comment by rtkaczyk** [comment]
Please add the new opcodes to `constGasFees` in `OpCodeGasSpec`. That should be used in line 104 (my mistake). Could you please also change that line to: 

```
verifyGas(constGasFees(op), stateIn, stateOut)
```

?

File: src/main/scala/io/iohk/ethereum/vm/DataWord.scala [hunk 7]
```diff
@@ -79,10 +88,56 @@ class DataWord private (private val n: BigInt) extends Ordered[DataWord] {
...
+    DataWord((this.n + that.n) mod modulus.n)
+  }
+
+  def mulmod(that: DataWord, modulus: DataWord): DataWord = zeroCheck(modulus) {
+    DataWord((this.n * that.n) mod modulus.n)
+  }
 
   def **(that: DataWord): DataWord = DataWord(this.n.modPow(that.n, Modulus))
 
+  def signExtend(that: DataWord): DataWord = {
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: Regarding @rtkaczyk said about the `.toByte` conversion being wrong in `getByte`, isn't it also the case here? Geth also

File: src/main/scala/io/iohk/ethereum/vm/DataWord.scala (hunk 2 above)
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: Also no need to perform bytes gymnastics or for `{}` in the conditional: 

```scala
    if (cnt > 31) Zero else DataWord(bytes(cnt))
```

Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: Also, no need for the conditional :grin::

```scala
def getByte(that: DataWord): DataWord =
  DataWord(n >> (256 - 8 * (that.n + 1)) & 0xff)
```

File: src/main/scala/io/iohk/ethereum/vm/OpCode.scala (hunk 1 above)
Comment (whysoserious) [inline src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: Note to myself: never write code during the meetings ;)

File: src/main/scala/io/iohk/ethereum/vm/DataWord.scala [hunk 8]
```diff
@@ -79,10 +88,56 @@ class DataWord private (private val n: BigInt) extends Ordered[DataWord] {
...
+  }
+
+  def mod(that: DataWord): DataWord = zeroCheck(that) { DataWord(this.n mod that.n) }
+
+  def smod(that: DataWord): DataWord = zeroCheck(that) {
+    DataWord(this.signedN.signum * (this.signedN.abs mod that.signedN.abs))
+  }
+
+  def addmod(that: DataWord, modulus: DataWord): DataWord = zeroCheck(modulus) {
+    DataWord((this.n + that.n) mod modulus.n)
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: It would be a bit faster to use `new DataWord(...)` instead of `DataWord(...)`. Since the resulting integer must be within bounds `[0, MaxValue]` there's no need to call `fixBigInt`.

This also pertains to other methods such as: / (div), sdiv, mod, smod, addmod, mulmod, **, signExtend, getByte, sgt, slt, and, or, xor

File: src/main/scala/io/iohk/ethereum/vm/DataWord.scala [hunk 9]
```diff
@@ -79,10 +88,56 @@ class DataWord private (private val n: BigInt) extends Ordered[DataWord] {
...
+  def /(that: DataWord): DataWord = zeroCheck(that) { DataWord(this.n / that.n) }
+
+  def sdiv(that: DataWord): DataWord = zeroCheck(that) {
+    DataWord(toUnsignedBigInt(this.signedN / that.signedN))
+  }
+
+  def mod(that: DataWord): DataWord = zeroCheck(that) { DataWord(this.n mod that.n) }
+
+  def smod(that: DataWord): DataWord = zeroCheck(that) {
+    DataWord(this.signedN.signum * (this.signedN.abs mod that.signedN.abs))
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: Isn't that equivalent to: `DataWord(this.signedN % that.signedN)` ?

File: src/main/scala/io/iohk/ethereum/vm/DataWord.scala [hunk 10]
```diff
@@ -79,10 +88,56 @@ class DataWord private (private val n: BigInt) extends Ordered[DataWord] {
...
   def *(that: DataWord): DataWord = DataWord(this.n * that.n)
 
-  def /(that: DataWord): DataWord = DataWord(this.n / that.n)
+  def /(that: DataWord): DataWord = zeroCheck(that) { DataWord(this.n / that.n) }
+
+  def sdiv(that: DataWord): DataWord = zeroCheck(that) {
+    DataWord(toUnsignedBigInt(this.signedN / that.signedN))
+  }
+
+  def mod(that: DataWord): DataWord = zeroCheck(that) { DataWord(this.n mod that.n) }
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: `%` and `mod` are the same for non-negative numbers but `%` is slightly faster.

File: src/main/scala/io/iohk/ethereum/vm/DataWord.scala [hunk 11]
```diff
@@ -79,10 +88,56 @@ class DataWord private (private val n: BigInt) extends Ordered[DataWord] {
...
 
+  def signExtend(that: DataWord): DataWord = {
+    val cnt: Byte = that.n.toByte
+    if (cnt < 0 || cnt > 31) {
+      this
+    } else {
+      val mask: Int = if (this.signedN.testBit((cnt * 8) + 7)) 0xFF else 0x00
+      val leftFill: Array[Byte] = Array.fill(Size - cnt - 1)(mask.toByte)
+      val bs = ByteString(leftFill) ++ this.bytes.takeRight(cnt + 1)
+      DataWord(bs)
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: Nothing wrong with this, but I would I take a different approach:

```scala
val negative = n.testBit(cnt * 8 + 7)
val mask = (BigInt(1) << ((cnt + 1) * 8)) - 1 // == BigInt(2).pow((cnt + 1) * 8) - 1
val newN = if (negative) n | (MaxValue ^ mask) else n & mask
new DataWord(newN)
```

I would expect this to be a tad faster (though not 100% sure)

File: src/test/scala/io/iohk/ethereum/vm/DataWordSpec.scala (hunk 6 above)
Comment (rtkaczyk) [inline src/test/scala/io/iohk/ethereum/vm/DataWordSpec.scala]: Also we could use more numbers like that, with 42 the result is always either 42 or 0.

File: src/main/scala/io/iohk/ethereum/vm/DataWord.scala [hunk 12]
```diff
@@ -79,10 +88,56 @@ class DataWord private (private val n: BigInt) extends Ordered[DataWord] {
...
+
+  def mulmod(that: DataWord, modulus: DataWord): DataWord = zeroCheck(modulus) {
+    DataWord((this.n * that.n) mod modulus.n)
+  }
 
   def **(that: DataWord): DataWord = DataWord(this.n.modPow(that.n, Modulus))
 
+  def signExtend(that: DataWord): DataWord = {
+    val cnt: Byte = that.n.toByte
+    if (cnt < 0 || cnt > 31) {
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: I don't like the name `cnt`. It suggests we're counting something, but we're not, it's a constant index. I would propose `i` or `idx` or `index`. Same goes for `getByte`.

File: src/main/scala/io/iohk/ethereum/vm/DataWord.scala [hunk 13]
```diff
@@ -79,10 +88,56 @@ class DataWord private (private val n: BigInt) extends Ordered[DataWord] {
...
+      Zero
+    } else {
+      DataWord(Zeros.take(Size - 1) :+ this.bytes(cnt))
+    }
+  }
+
+  private def zeroCheck(dw: DataWord)(result: =>DataWord): DataWord =
+    if (dw == Zero) Zero else result
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: Neat :+1:

File: src/test/scala/io/iohk/ethereum/vm/DataWordSpec.scala [hunk 14]
```diff
@@ -100,9 +106,57 @@ class DataWordSpec extends FunSuite with PropertyChecks {
...
+      }
     }
+    assert((DataWord(-1) mod DataWord(MaxValue.toBigInt)) == Zero)
+    assert((DataWord(1) mod Zero) == Zero)
+  }
+
+  test("smod") {
+    assert((DataWord(Modulus - 1) smod DataWord(3)) == DataWord(Modulus - 1))
+    assert((DataWord(-1) smod DataWord(MaxValue.toBigInt)) == Zero)
+    assert((DataWord(1) smod Zero) == Zero)
```
Comment (rtkaczyk) [inline src/test/scala/io/iohk/ethereum/vm/DataWordSpec.scala]: `DataWord(Modulus - 1` and `DataWord(MaxValue.toBigInt)` are simply equivalent to `MaxValue`.
Also this could test more cases.

File: src/main/scala/io/iohk/ethereum/vm/DataWord.scala (hunk 4 above)
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: If comparisons return `DataWord` it will be more convenient for the opcodes' implementation.

On the other hand, I was thinking that `vm.DataWord` could evolve into `domain.Uint256` and `domain.Int256`, given that such numbers are used across the whole system. For that reason `Uint256` should fully comply with Scala conventions for numeric types.

@adamsmo you researched the numbers a bit, would such numeric types be useful for us?
@mcsherrylabs any thoughts on this?

Comment (whysoserious) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: > On the other hand, I was thinking that vm.DataWord could evolve into domain.Uint256 and domain.Int256

I would keep it as it is now. Most important information for the rest of the system is that a `DataWord` is a 32 bytes long array.

Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: @rtkaczyk I think yes it would be helpful to have something like this, but we have to provide conversion to bytes for that type (for RLP encoding)

Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: > Most important information for the rest of the system is that a DataWord is a 32 bytes long array.

I disagree about that. It should be regarded as a number `P_256`, that can also be viewed as a byte array. Our usages in the vm code reflect that, in that it is much more often treated as a number, rather than as a byte array.

Comment (whysoserious) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: Absoulely true. Now I agree with you that this is a good idea.

Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: Also parity uses something like `U256` all around. Anyway, I think changing `DataWord` in this manner is outside the scope of this PR, so maybe let's just leave it as is, and provide a comment like:

```
// TODO: consider moving to domain as Uint256, which follows Scala numeric conventions, and is used across the system for P_256 numbers (see YP 4.3)
```

**Comment by whysoserious** [comment]
@ntallar, @rtkaczyk I think that this PR is ready for review again.

I didn't address all of your comments re. `DataWord` as I am planning to split it up into two new types: `UInt256` and `SInt256`. I assume that `DataWord.scala` and `DataWordSpec.scala` will be completely rewritten from scratch in an upcoming PR.

**Comment by rtkaczyk** [comment]
> I assume that DataWord.scala and DataWordSpec.scala will be completely rewritten from scratch

Only if you are going to change the underlying implementation. I wouldn't recommend that as the first step. I'd rather we use those new types right away. And in that case we can reuse most of this code.

Also, I wouldn't call it SInt. I believe all the languages that differentiate signed and unsigned would call them Int and Uint.

File: src/main/scala/io/iohk/ethereum/vm/DataWord.scala (hunk 12 above)
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: We should still at least achieve correctness. That means `that.n.toByte` has to go.

Comment (whysoserious) [inline src/main/scala/io/iohk/ethereum/vm/DataWord.scala]: Argh, right. This somehow slipped through my own review. Fixed.
